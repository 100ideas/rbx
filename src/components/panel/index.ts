import "./panel.sass";

export { default } from "./Panel";
