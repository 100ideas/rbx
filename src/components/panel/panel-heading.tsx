import classNames from "classNames";
import * as React from "react";

import { forwardRefAs, Generic, HelpersProps } from "src/base";

export type PanelHeadingProps = HelpersProps;

export const PanelHeading = forwardRefAs<PanelHeadingProps, "div">(
  ({ className, ...rest }, ref) => (
    <Generic
      className={classNames("panel-heading", className)}
      ref={ref}
      {...rest}
    />
  ),
  { as: "div" },
);
