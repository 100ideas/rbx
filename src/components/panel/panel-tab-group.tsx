import classNames from "classnames";
import React from "react";

import { forwardRefAs, Generic, HelpersProps } from "../../base";

export type PanelTabGroupProps = HelpersProps;

export const PanelTabGroup = forwardRefAs<PanelTabGroupProps, "div">(
  ({ className, ...rest }, ref) => (
    <Generic
      className={classNames("panel-tabs", className)}
      ref={ref}
      {...rest}
    />
  ),
  { as: "div" },
);

PanelTabGroup.displayName = "Panel.Tab.Group";
