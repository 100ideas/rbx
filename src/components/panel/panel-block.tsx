import { cx } from "emotion";
import React from "react";

import { asExoticComponent } from "@/components/exotic";
import { ModifierProps, modify } from "@/modifiers";

export type PanelBlockModifierProps = Partial<{
  active: boolean;
}>;

export type PanelBlockProps = ModifierProps & PanelBlockModifierProps;

export const PanelBlock = asExoticComponent<PanelBlockProps, "div">(
  (props, ref) => {
    const { active, as, ...rest } = modify(props);
    rest.className = cx("panel-block", rest.className, {
      "is-active": active,
    });
    return React.createElement(as!, { ref, ...rest });
  },
  "div",
);

PanelBlock.defaultProps = Object.assign(
  { active: false },
  PanelBlock.defaultProps,
);
