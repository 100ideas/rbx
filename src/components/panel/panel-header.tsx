import { cx } from "emotion";
import React from "react";

import { asExoticComponent } from "@/components/exotic";
import { ModifierProps, modify } from "@/modifiers";

export type PanelHeaderProps = ModifierProps;

export const PanelHeader = asExoticComponent<PanelHeaderProps, "div">(
  (props, ref) => {
    const { as, ...rest } = modify(props);
    rest.className = cx("panel-heading", rest.className);
    return React.createElement(as!, { ref, ...rest });
  },
  "div",
);
