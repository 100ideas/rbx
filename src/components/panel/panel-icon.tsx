import { cx } from "emotion";
import React from "react";

import { asExoticComponent } from "@/components/exotic";
import { ModifierProps, modify } from "@/modifiers";

export type PanelIconProps = ModifierProps;

export const PanelIcon = asExoticComponent<PanelIconProps, "span">(
  (props, ref) => {
    const { as, ...rest } = modify(props);
    rest.className = cx("panel-icon", rest.className);
    return React.createElement(as!, { ref, ...rest });
  },
  "span",
);
