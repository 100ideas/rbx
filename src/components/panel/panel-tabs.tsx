import { cx } from "emotion";
import React from "react";

import { Element, extendedForwardRef } from "@/components/element";
import { ModifierProps } from "@/modifiers";
import { PanelTabsTab } from "./panel-tabs-tab";

export type PanelTabsProps = ModifierProps;

export const PanelTabs = Object.assign(
  extendedForwardRef<PanelTabsProps, "div">(
    ({ className, ...props }, ref) => (
      <Element {...props} ref={ref} className={cx("panel-tabs", className)} />
    ),
    "div",
  ),
  { Tab: PanelTabsTab },
);
