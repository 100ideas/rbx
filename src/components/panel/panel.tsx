import { cx } from "emotion";
import React from "react";

import { asExoticComponent } from "@/components/exotic";
import { ModifierProps, transformModifiers } from "@/modifiers";
import { PanelBlock } from "./panel-block";
import { PanelHeader } from "./panel-header";
import { PanelIcon } from "./panel-icon";
import { PanelTabs } from "./panel-tabs";

export type PanelProps = ModifierProps;

export const Panel = Object.assign(
  asExoticComponent<PanelProps, "nav">((props, ref) => {
    const { as, ...rest } = transformModifiers(props);
    rest.className = cx("panel", rest.className);
    return React.createElement(as!, { ref, ...rest });
  }, "nav"),
  {
    Block: PanelBlock,
    Header: PanelHeader,
    Icon: PanelIcon,
    Tabs: PanelTabs,
  },
);
