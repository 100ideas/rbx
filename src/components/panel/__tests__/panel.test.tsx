import { Panel } from "../panel";
import { PanelBlock } from "../panel-block";
import { PanelHeading } from "../panel-heading";
import { PanelIcon } from "../panel-icon";
import { PanelTabs } from "../panel-tabs";

import {
  hasProperties,
  makeNodeFactory,
  makeShallowWrapper,
  testForwardRefAsExoticComponentIntegration,
  testThemeIntegration,
} from "../../../__tests__/testing";

const COMPONENT = Panel;
const COMPONENT_NAME = "Panel";
const DEFAULT_ELEMENT = "nav";
const BULMA_CLASS_NAME = "panel";

const makeNode = makeNodeFactory(COMPONENT);

describe(`${COMPONENT_NAME} component`, () => {
  hasProperties(COMPONENT, {
    Block: PanelBlock,
    Heading: PanelHeading,
    Icon: PanelIcon,
    Tabs: PanelTabs,
    defaultProps: { as: DEFAULT_ELEMENT },
  });

  testForwardRefAsExoticComponentIntegration(
    makeNode,
    makeShallowWrapper,
    DEFAULT_ELEMENT,
    BULMA_CLASS_NAME,
  );

  testThemeIntegration(makeNode, makeShallowWrapper);
});
