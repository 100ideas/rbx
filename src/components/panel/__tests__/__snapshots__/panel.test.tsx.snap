// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel component should Exist 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "Block": Object {
    "$$typeof": Symbol(react.forward_ref),
    "defaultProps": Object {
      "active": false,
      "as": "div",
    },
    "render": [Function],
  },
  "Header": Object {
    "$$typeof": Symbol(react.forward_ref),
    "defaultProps": Object {
      "as": "div",
    },
    "render": [Function],
  },
  "Icon": Object {
    "$$typeof": Symbol(react.forward_ref),
    "defaultProps": Object {
      "as": "span",
    },
    "render": [Function],
  },
  "Tabs": Object {
    "$$typeof": Symbol(react.forward_ref),
    "Tab": Object {
      "$$typeof": Symbol(react.forward_ref),
      "defaultProps": Object {
        "active": false,
        "as": "a",
      },
      "render": [Function],
    },
    "defaultProps": Object {
      "as": "div",
    },
    "render": [Function],
  },
  "defaultProps": Object {
    "as": "nav",
  },
  "render": [Function],
}
`;

exports[`Panel component should have box classname 1`] = `
<nav
  className="panel"
>
  <div
    className="panel-heading"
  >
    repositories
  </div>
  <div
    className="panel-block"
  >
    <div>
      Control
    </div>
  </div>
  <div
    className="panel-tabs panel-tabs"
  >
    <a
      className="is-active"
    >
      all
    </a>
    <a>
      public
    </a>
    <a>
      private
    </a>
    <a>
      sources
    </a>
    <a>
      forks
    </a>
  </div>
  <a
    className="panel-block is-active"
  >
    <span
      className="panel-icon"
    >
      <i
        className="fa fa-bars"
      />
    </span>
    bulma
  </a>
  <label
    className="panel-block panel-block"
  >
    <input
      type="checkbox"
    />
    remember me
  </label>
  <div
    className="panel-block"
  >
    <button
      type="button"
    >
      reset all filters
    </button>
  </div>
</nav>
`;
